src/goto-symex/goto_symex.cpp

```diff
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
     if(state.source.pc->source_location.get_hide())
       assignment_type = symex_targett::assignment_typet::HIDDEN;
 
+    symex_assignt symex_assign{
+      state, assignment_type, ns, symex_config, target};
+
+    if(constant_propagate_assignment_with_side_effects(
+         state, symex_assign, lhs, rhs))
+      return;
+
     exprt::operandst lhs_if_then_else_conditions;
-    symex_assignt{state, assignment_type, ns, symex_config, target}.assign_rec(
+    symex_assign.assign_rec(
       lhs, expr_skeletont{}, rhs, lhs_if_then_else_conditions);
   }
 }
+
+bool goto_symext::constant_propagate_assignment_with_side_effects(
+  statet &state,
+  symex_assignt &symex_assign,
+  const exprt &lhs,
+  const exprt &rhs)
+{
+  if(!symex_config.simplify_opt)
+  {
+    return false;
+  }
+
+  if(state.threads.size() > 1)
+  {
+    return false;
+  }
+
+  if(rhs.id() == ID_function_application)
+  {
+    exprt rhs_l2 = state.rename(rhs, ns).get();
```

Comment:
@danpoe I don't really understand why we do an L2 renaming here. Can we look at this together please?

Reply:
Ah, I initially did this since `l2` renaming also replaces `l1` symbols with `l2` expressions stored in the propagation map apparently. But I think I can retrieve the expressions from the propagation map directly instead of doing the renaming.

Reply:
@danpoe, so are you doing this?

Reply:
Ah yes this code can now be removed.